pallets/foreign-investments/src/impls/mod.rs

```diff
@@ -0,0 +1,1152 @@
+			})?;
+		Pallet::<T>::apply_redeem_state_transition(who, investment_id, post_state)?;
+
+		Ok(amount)
+	}
+
+	#[transactional]
+	fn collect_foreign_investment(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+		foreign_payout_currency: T::CurrencyId,
+		pool_currency: T::CurrencyId,
+	) -> Result<ExecutedForeignCollectInvest<T::Balance>, DispatchError> {
+		// Note: We assume the configured Investment trait to notify about the collected
+		// amounts via the `CollectedInvestmentHook` which handles incrementing the
+		// `CollectedInvestment` amount.
+		T::Investment::collect_investment(who.clone(), investment_id)?;
+
+		Self::transfer_collected_investment(
+			who,
+			investment_id,
+			foreign_payout_currency,
+			pool_currency,
+		)
+	}
+
+	#[transactional]
+	fn collect_foreign_redemption(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+		foreign_payout_currency: T::CurrencyId,
+		pool_currency: T::CurrencyId,
+	) -> Result<(), DispatchError> {
+		ensure!(
+			RedemptionPayoutCurrency::<T>::get(who, investment_id)
+				.map(|currency| currency == foreign_payout_currency)
+				.unwrap_or_else(|| {
+					log::debug!("Corruption: Redemption payout currency missing when calling decrease. Should never occur if redemption has been increased beforehand");
+					false
+				}),
+			Error::<T>::InvalidRedemptionPayoutCurrency
+		);
+		ensure!(T::PoolInspect::currency_for(investment_id.of_pool())
+			.map(|currency| currency == pool_currency)
+			.unwrap_or_else(|| {
+				log::debug!("Corruption: Failed to derive pool currency from investment id when collecting foreign redemption. Should never occur if redemption has been increased beforehand");
+				false
+			}),
+			DispatchError::Corruption
+		);
+
+		// Note: We assume the configured Investment trait to notify about the collected
+		// amounts via the `CollectedRedemptionHook` which handles incrementing the
+		// `CollectedRedemption` amount.
+		T::Investment::collect_redemption(who.clone(), investment_id)?;
+
+		Ok(())
+	}
+
+	fn investment(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+	) -> Result<T::Balance, DispatchError> {
+		T::Investment::investment(who, investment_id)
+	}
+
+	fn redemption(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+	) -> Result<T::Balance, DispatchError> {
+		T::Investment::redemption(who, investment_id)
+	}
+
+	fn accepted_payment_currency(investment_id: T::InvestmentId, currency: T::CurrencyId) -> bool {
+		if T::Investment::accepted_payment_currency(investment_id, currency) {
+			true
+		} else {
+			T::PoolInspect::currency_for(investment_id.of_pool())
+				.map(|pool_currency| {
+					// TODO(@review): Or just `order_pair_exists`?
+					T::TokenSwaps::counter_order_pair_exists(currency, pool_currency)
+				})
+				.unwrap_or(false)
+		}
+	}
+
+	fn accepted_payout_currency(investment_id: T::InvestmentId, currency: T::CurrencyId) -> bool {
+		Self::accepted_payment_currency(investment_id, currency)
```

Comment:
Why call the `Self` method instead of calling `Investment::accepted_payout_currency`?

Reply:
This will be refactored after #1518:

```rust
	fn accepted_payout_currency(investment_id: T::InvestmentId, currency: T::CurrencyId) -> bool {
		if T::Investment::accepted_payout_currency(investment_id, currency) {
			true
		} else {
			T::PoolInspect::currency_for(investment_id.of_pool())
				.map(|pool_currency| T::TokenSwaps::valid_pair(pool_currency, currency))
				.unwrap_or(false)
		}
	}
```

Reply:
Ok, gotcha! 👍🏻 
Thanks for the code explanation